File: BST/simple_bst/showlibia/Java-version/Correct.java
package bst;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;


public class Correct {

    public static void main(String[] args) {
        String inputFilename = "../../../benchmark/benchmark_operations.csv";  
        String outputFilename = "correctness.csv";           

        BST bst = new BST();

        try (BufferedReader br = new BufferedReader(new FileReader(inputFilename));
             BufferedWriter bw = new BufferedWriter(new FileWriter(outputFilename))) {

            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length == 2 || parts.length == 3) {
                    String operation = parts[0];
                    int value = Integer.parseInt(parts[1]);
                    processOperation(bst, operation, value, bw);
                } else {
                    System.err.println("Warning: Malformed line: " + line);
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void processOperation(BST bst, String operation, int value, BufferedWriter bw) throws IOException {
        if (operation.equals("insert")) {
            bst.insert(value);
            bw.write("insert," + value + "\n");
        } else if (operation.equals("search")) {
            boolean found = bst.contains(value);
            bw.write("search," + value + "," + (found ? "true" : "false") + "\n");
        } else if (operation.equals("delete")) {
            boolean found = bst.contains(value);
            bst.remove(value);
            bw.write("delete," + value + (found ? "" : ",null") + "\n");
        } else if (operation.equals("findmin")) {
            BSTNode minNode = bst.findMin();
            if (minNode != null) {
                bw.write("findmin," + minNode.data + "\n");
            } else {
                bw.write("findmin,null\n");
            }
        } else if (operation.equals("findmax")) {
            BSTNode maxNode = bst.findMax();
            if (maxNode != null) {
                bw.write("findmax," + maxNode.data + "\n");
            } else {
                bw.write("findmax,null\n");
            }
        } else {
            System.err.println("Warning: Unknown operation: " + operation);
        }
    }
}

File: BST/simple_bst/showlibia/Cpp-version/correct.cc
#include "bst.h" // Assuming the C++ BST implementation is in bst.h
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct Operation {
  string operation;
  int value;
};

int load_operations(std::string_view filename, vector<Operation> &operations) {
  ifstream file((string)filename);
  if (!file.is_open()) {
    cerr << "Error opening file: " << filename << endl;
    return -1;
  }

  string line;
  while (getline(file, line)) {
    istringstream iss(line);
    string op;
    int value;

    if (getline(iss, op, ',') && iss >> value) {
      operations.push_back({op, value});
    } else {
      cerr << "Warning: Malformed line: " << line << endl;
    }
  }

  file.close();
  return operations.size();
}

void process_operation(BST &bst, const string &operation, int value,
                       ofstream &output_file) {
  if (operation == "insert") {
    bst.insert(value);
    output_file << "insert," << value << endl;
  } else if (operation == "search") {
      bool found = bst.search(value);
      output_file << "search," << value << "," << (found ? "true" : "false")
                  << endl;
  } else if (operation == "delete") {
    bool found = bst.search(value);
    bst.remove(value);
    output_file << "delete," << value << (found ? "" : ",null") << endl;
  } else if (operation == "findmin") {
    BSTNode *min_node = bst.findMin();
    if (min_node) {
      output_file << "findmin," << min_node->data << endl;
    } else {
      output_file << "findmin,null" << endl;
    }
  } else if (operation == "findmax") {
    BSTNode *max_node = bst.findMax();
    if (max_node) {
      output_file << "findmax," << max_node->data << endl;
    } else {
      output_file << "findmax,null" << endl;
    }
  } else {
    cerr << "Warning: Unknown operation: " << operation << endl;
  }
}

void execute_operations(const vector<Operation> &operations,
                        const string &output_filename) {
  ofstream output_file(output_filename);
  if (!output_file.is_open()) {
    cerr << "Error opening output file: " << output_filename << endl;
    return;
  }

  BST bst;

  for (const auto &op : operations) {
    process_operation(bst, op.operation, op.value, output_file);
  }

  output_file.close();
}

int main() {
  string input_filename = "../../../benchmark/benchmark_operations.csv";
  string output_filename = "correctness.csv";

  std::vector<Operation> operations;
  int count = load_operations(input_filename, operations);
  if (count < 0) {
    return 1;
  }

  execute_operations(operations, output_filename);

  return 0;
}